File: server/src/controllers/userController.ts
import { Request, Response } from 'express';
import { query } from '../config/db.js';
import { v4 as uuidv4 } from 'uuid';
import { normalizeString, toMySQLDateTime, buildSearchClause, buildSortClause } from '../utils/dataHelpers.js';
import bcrypt from 'bcryptjs';
import { AuthRequest } from '../middleware/auth.js';
import { getAvatarUrl, handleFileUpload } from '../middleware/upload.js';
import path from 'path';

export const getUsers = async (req: Request, res: Response) => {
  try {
    const { pageSize, pageIndex, search, sortBy, sortOrder } = req.query;
    
    // Parse pagination params with defaults
    const pageSizeNum = pageSize ? parseInt(pageSize as string, 10) : 10;
    const pageIndexNum = pageIndex ? parseInt(pageIndex as string, 10) : 0;
    const offset = pageIndexNum * pageSizeNum;
    
    // Build search and sort clauses
    const queryParams: any[] = [];
    const searchClause = buildSearchClause(
      search as string,
      ['name', 'email', 'phone', 'role'],
      queryParams
    );
    
    const allowedSortFields = ['name', 'email', 'phone', 'role', 'status', 'created_at', 'updated_at'];
    const sortClause = buildSortClause(sortBy as string, allowedSortFields, 'created_at', sortOrder as string);
    
    // Get total count with search
    // Update search clause to use table alias for count query
    const countSearchClause = searchClause
      .replace(/\bname\b/g, 'u.name')
      .replace(/\bemail\b/g, 'u.email')
      .replace(/\bphone\b/g, 'u.phone')
      .replace(/\brole\b/g, 'u.role');
    
    const countQuery = `SELECT COUNT(*) as total FROM users u ${countSearchClause}`;
    const countResults = await query<any[]>(countQuery, queryParams);
    const total = countResults[0]?.total || 0;
    
    // Get paginated data with role description
    // Update search and sort clauses to use table aliases
    const searchClauseWithAlias = searchClause
      .replace(/\bname\b/g, 'u.name')
      .replace(/\bemail\b/g, 'u.email')
      .replace(/\bphone\b/g, 'u.phone')
      .replace(/\brole\b/g, 'u.role');
    
    const sortClauseWithAlias = sortClause
      .replace(/\bname\b/g, 'u.name')
      .replace(/\bemail\b/g, 'u.email')
      .replace(/\bphone\b/g, 'u.phone')
      .replace(/\brole\b/g, 'u.role')
      .replace(/\bstatus\b/g, 'u.status')
      .replace(/\bcreated_at\b/g, 'u.created_at')
      .replace(/\bupdated_at\b/g, 'u.updated_at');
    
    const results = await query<any[]>(
      `SELECT 
        u.id, u.name, u.email, u.phone, 
        u.role,
        COALESCE(r.description, r.name) as role_description,
        u.status, u.avatar, u.created_at, u.updated_at 
      FROM users u
      LEFT JOIN roles r ON u.role = r.id
      ${searchClauseWithAlias} ${sortClauseWithAlias} LIMIT ${pageSizeNum} OFFSET ${offset}`,
      queryParams
    );
    
    res.json({
      data: results,
      total,
      pageIndex: pageIndexNum,
      pageSize: pageSizeNum,
    });
  } catch (error: any) {
    console.error('Error fetching users:', error);
    res.status(500).json({ error: 'Không thể lấy danh sách người dùng' });
  }
};

export const getUserById = async (req: Request, res: Response) => {
  try {
    const { id } = req.params;
    const results = await query<any[]>(
      `SELECT 
        u.id, u.name, u.email, u.phone, 
        u.role,
        COALESCE(r.description, r.name) as role_description,
        u.status, u.avatar, u.created_at, u.updated_at 
      FROM users u
      LEFT JOIN roles r ON u.role = r.id
      WHERE u.id = ?`,
      [id]
    );
    
    if (results.length === 0) {
      return res.status(404).json({ error: 'Không tìm thấy người dùng' });
    }
    
    res.json(results[0]);
  } catch (error: any) {
    console.error('Error fetching user:', error);
    res.status(500).json({ error: 'Không thể lấy thông tin người dùng' });
  }
};

export const createUser = async (req: Request, res: Response) => {
  try {
    const userData = req.body;
    const id = uuidv4();
    const createdAt = toMySQLDateTime();
    
    // Hash password if provided
    let passwordHash = userData.password_hash;
    if (userData.password && !passwordHash) {
      const saltRounds = 10;
      passwordHash = await bcrypt.hash(userData.password, saltRounds);
    }
    
    if (!passwordHash) {
      return res.status(400).json({ error: 'Vui lòng nhập mật khẩu' });
    }
    
    // Validate role_id
    let roleId = userData.role;
    if (!roleId) {
      // Try to get a default role (first role in database)
      try {
        const defaultRole = await query<any[]>(
          'SELECT id FROM roles LIMIT 1'
        );
        if (defaultRole.length > 0) {
          roleId = defaultRole[0].id;
        } else {
          return res.status(400).json({ error: 'No roles found in database. Please create at least one role first.' });
        }
      } catch (roleError) {
        return res.status(400).json({ error: 'Role (role_id) is required and no default role found' });
      }
    } else {
      // Validate that the role exists
      try {
        const roleCheck = await query<any[]>(
          'SELECT id FROM roles WHERE id = ?',
          [roleId]
        );
        if (roleCheck.length === 0) {
          return res.status(400).json({ error: `Role with id ${roleId} does not exist` });
        }
      } catch (roleError: any) {
        return res.status(400).json({ error: `Invalid role_id: ${roleError.message}` });
      }
    }
    
    await query(
      `INSERT INTO users (
        id, name, email, phone, password_hash, role, status, avatar, created_at, updated_at
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      [
        id,
        userData.name,
        userData.email,
        userData.phone || '',
        passwordHash,
        roleId,
        userData.status || 'active',
        normalizeString(userData.avatar),
        createdAt,
        createdAt,
      ]
    );
    
    const newUser = await query<any[]>(
      `SELECT 
        u.id, u.name, u.email, u.phone, 
        u.role,
        COALESCE(r.description, r.name) as role_description,
        u.status, u.avatar, u.created_at, u.updated_at 
      FROM users u
      LEFT JOIN roles r ON u.role = r.id
      WHERE u.id = ?`,
      [id]
    );
    
    res.status(201).json(newUser[0]);
  } catch (error: any) {
    console.error('Error creating user:', error);
    console.error('Request data:', req.body);
    
    // Handle duplicate email error
    if (error.code === 'ER_DUP_ENTRY' || error.message?.includes('Duplicate entry')) {
      return res.status(400).json({ error: 'Email đã tồn tại' });
    }
    
    res.status(500).json({ 
      error: 'Không thể tạo người dùng',
      message: error.message 
    });
  }
};

export const updateUser = async (req: Request, res: Response) => {
  try {
    const { id } = req.params;
    const userData = req.body;
    const updatedAt = toMySQLDateTime();
    
    // Check if exists
    const existing = await query<any[]>(
      'SELECT * FROM users WHERE id = ?',
      [id]
    );
    
    if (existing.length === 0) {
      return res.status(404).json({ error: 'Không tìm thấy người dùng' });
    }
    
    // Hash password if provided
    let passwordHash = userData.password_hash;
    if (userData.password && !passwordHash) {
      const saltRounds = 10;
      passwordHash = await bcrypt.hash(userData.password, saltRounds);
    }
    
    // Build update query dynamically
    const updates: string[] = [];
    const values: any[] = [];
    
    if (userData.name !== undefined) {
      updates.push('name = ?');
      values.push(userData.name);
    }
    if (userData.email !== undefined) {
      updates.push('email = ?');
      values.push(userData.email);
    }
    if (userData.phone !== undefined) {
      updates.push('phone = ?');
      values.push(userData.phone);
    }
    if (passwordHash) {
      updates.push('password_hash = ?');
      values.push(passwordHash);
    }
    if (userData.role !== undefined) {
      // Validate that the role exists
      try {
        const roleCheck = await query<any[]>(
          'SELECT id FROM roles WHERE id = ?',
          [userData.role]
        );
        if (roleCheck.length === 0) {
          return res.status(400).json({ error: `Role with id ${userData.role} does not exist` });
        }
      updates.push('role = ?');
      values.push(userData.role);
      } catch (roleError: any) {
        return res.status(400).json({ error: `Invalid role_id: ${roleError.message}` });
      }
    }
    if (userData.status !== undefined) {
      updates.push('status = ?');
      values.push(userData.status);
    }
    if (userData.avatar !== undefined) {
      updates.push('avatar = ?');
      values.push(normalizeString(userData.avatar));
    }
    
    updates.push('updated_at = ?');
    values.push(updatedAt);
    values.push(id);
    
    await query(
      `UPDATE users SET ${updates.join(', ')} WHERE id = ?`,
      values
    );
    
    const updated = await query<any[]>(
      `SELECT 
        u.id, u.name, u.email, u.phone, 
        u.role,
        COALESCE(r.description, r.name) as role_description,
        u.status, u.avatar, u.created_at, u.updated_at 
      FROM users u
      LEFT JOIN roles r ON u.role = r.id
      WHERE u.id = ?`,
      [id]
    );
    
    res.json(updated[0]);
  } catch (error: any) {
    console.error('Error updating user:', error);
    console.error('Request data:', req.body);
    
    // Handle duplicate email error
    if (error.code === 'ER_DUP_ENTRY' || error.message?.includes('Duplicate entry')) {
      return res.status(400).json({ error: 'Email đã tồn tại' });
    }
    
    res.status(500).json({ 
      error: 'Không thể cập nhật người dùng',
      message: error.message 
    });
  }
};

export const deleteUser = async (req: Request, res: Response) => {
  try {
    const { id } = req.params;
    
    // Check if exists
    const existing = await query<any[]>(
      'SELECT * FROM users WHERE id = ?',
      [id]
    );
    
    if (existing.length === 0) {
      return res.status(404).json({ error: 'Không tìm thấy người dùng' });
    }
    
    await query('DELETE FROM users WHERE id = ?', [id]);
    
    res.status(204).send();
  } catch (error: any) {
    console.error('Error deleting user:', error);
    res.status(500).json({ error: 'Không thể xóa người dùng' });
  }
};

// Get current user (from JWT token)
export const getCurrentUser = async (req: AuthRequest, res: Response) => {
  try {
    if (!req.userId) {
      return res.status(401).json({ error: 'Không có quyền truy cập' });
    }
    
    const results = await query<any[]>(
      `SELECT 
        u.id, u.name, u.email, u.phone, 
        u.role,
        COALESCE(r.description, r.name) as role_description,
        u.status, u.avatar, u.created_at, u.updated_at 
      FROM users u
      LEFT JOIN roles r ON u.role = r.id
      WHERE u.id = ?`,
      [req.userId]
    );
    
    if (results.length === 0) {
      return res.status(404).json({ error: 'Không tìm thấy người dùng' });
    }
    
    const user = results[0];
    // Convert avatar filename to URL if exists
    if (user.avatar) {
      user.avatar = getAvatarUrl(user.avatar);
    }
    
    res.json(user);
  } catch (error: any) {
    console.error('Error fetching current user:', error);
    res.status(500).json({ error: 'Không thể lấy thông tin người dùng hiện tại' });
  }
};

// Upload avatar
export const uploadAvatar = async (req: AuthRequest, res: Response) => {
  try {
    if (!req.userId) {
      return res.status(401).json({ error: 'Không có quyền truy cập' });
    }

    if (!req.file) {
      return res.status(400).json({ error: 'Không có file được tải lên' });
    }

    const updatedAt = toMySQLDateTime();

    // Handle file upload (Supabase or filesystem)
    const { filename, url } = await handleFileUpload(req.file, 'avatars');
    
    // Store filename or full URL in database depending on storage type
    // If URL is a full URL (Supabase), store the full URL, otherwise store filename
    const avatarValue = url.startsWith('http') ? url : filename;

    // Update user avatar in database
    await query(
      'UPDATE users SET avatar = ?, updated_at = ? WHERE id = ?',
      [avatarValue, updatedAt, req.userId]
    );

    // Get updated user
    const results = await query<any[]>(
      `SELECT 
        u.id, u.name, u.email, u.phone, 
        u.role,
        COALESCE(r.description, r.name) as role_description,
        u.status, u.avatar, u.created_at, u.updated_at 
      FROM users u
      LEFT JOIN roles r ON u.role = r.id
      WHERE u.id = ?`,
      [req.userId]
    );

    if (results.length === 0) {
      return res.status(404).json({ error: 'Không tìm thấy người dùng' });
    }

    const user = results[0];
    user.avatar = getAvatarUrl(user.avatar);

    res.json({
      message: 'Tải lên ảnh đại diện thành công',
      avatar: user.avatar,
      user,
    });
  } catch (error: any) {
    console.error('Error uploading avatar:', error);
    res.status(500).json({ 
      error: 'Không thể tải lên ảnh đại diện',
      message: error.message 
    });
  }
};

File: server.net/Controllers/UserController.cs
using ConstructOs.Server.Config;
using ConstructOs.Server.Middleware;
using ConstructOs.Server.Models;
using ConstructOs.Server.Utils;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using BCrypt.Net;

namespace ConstructOs.Server.Controllers;

[ApiController]
[Route("api/users")]
public class UserController : ControllerBase
{
    private readonly IDatabaseService _db;
    private readonly IConfiguration _configuration;
    private readonly ILogger<UserController> _logger;
    private readonly IWebHostEnvironment _environment;

    public UserController(
        IDatabaseService db, 
        IConfiguration configuration, 
        ILogger<UserController> logger,
        IWebHostEnvironment environment)
    {
        _db = db;
        _configuration = configuration;
        _logger = logger;
        _environment = environment;
    }

    [HttpGet]
    public async Task<IActionResult> GetUsers([FromQuery] int? pageSize, [FromQuery] int? pageIndex, 
        [FromQuery] string? search, [FromQuery] string? sortBy, [FromQuery] string? sortOrder)
    {
        try
        {
            var pageSizeNum = pageSize ?? 10;
            var pageIndexNum = pageIndex ?? 0;
            var offset = pageIndexNum * pageSizeNum;

            var queryParams = new List<object>();
            var searchClause = DataHelpers.BuildSearchClause(search, 
                new[] { "name", "email", "phone", "role" }, out queryParams);

            var countSearchClause = searchClause
                .Replace("name", "u.name")
                .Replace("email", "u.email")
                .Replace("phone", "u.phone")
                .Replace("role", "u.role");

            var countSql = $"SELECT COUNT(*) as total FROM users u {countSearchClause}";
            dynamic? countResult = null;
            if (queryParams.Count > 0)
            {
                // Pass parameters as array for Dapper
                countResult = await _db.QueryFirstOrDefaultAsync<dynamic>(countSql, queryParams.ToArray());
            }
            else
            {
                countResult = await _db.QueryFirstOrDefaultAsync<dynamic>(countSql);
            }
            var total = Convert.ToInt32(countResult?.total ?? 0);

            var allowedSortFields = new[] { "name", "email", "phone", "role", "status", "created_at", "updated_at" };
            var sortClause = DataHelpers.BuildSortClause(sortBy, allowedSortFields, "created_at", sortOrder);
            var sortClauseWithAlias = sortClause
                .Replace("name", "u.name")
                .Replace("email", "u.email")
                .Replace("phone", "u.phone")
                .Replace("role", "u.role")
                .Replace("status", "u.status")
                .Replace("created_at", "u.created_at")
                .Replace("updated_at", "u.updated_at");

            var searchClauseWithAlias = searchClause
                .Replace("name", "u.name")
                .Replace("email", "u.email")
                .Replace("phone", "u.phone")
                .Replace("role", "u.role");

            var sql = $@"SELECT 
                u.id, u.name, u.email, u.phone, 
                u.role,
                COALESCE(r.description, r.name) as role_description,
                u.status, u.avatar, u.created_at, u.updated_at 
            FROM users u
            LEFT JOIN roles r ON u.role = r.id
            {searchClauseWithAlias} {sortClauseWithAlias} LIMIT {pageSizeNum} OFFSET {offset}";

            IEnumerable<dynamic> results;
            if (queryParams.Count > 0)
            {
                // Pass parameters as array for Dapper
                results = await _db.QueryAsync<dynamic>(sql, queryParams.ToArray());
            }
            else
            {
                results = await _db.QueryAsync<dynamic>(sql);
            }

            return Ok(new
            {
                data = results,
                total,
                pageIndex = pageIndexNum,
                pageSize = pageSizeNum
            });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching users");
            return StatusCode(500, new { error = "Không thể lấy danh sách người dùng" });
        }
    }

    [HttpGet("{id}")]
    public async Task<IActionResult> GetUserById(string id)
    {
        try
        {
            var sql = @"SELECT 
                u.id, u.name, u.email, u.phone, 
                u.role,
                COALESCE(r.description, r.name) as role_description,
                u.status, u.avatar, u.created_at, u.updated_at 
            FROM users u
            LEFT JOIN roles r ON u.role = r.id
            WHERE u.id = ?";

            var user = await _db.QueryFirstOrDefaultAsync<dynamic>(sql, new { id });

            if (user == null)
            {
                return NotFound(new { error = "Không tìm thấy người dùng" });
            }

            return Ok(user);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching user");
            return StatusCode(500, new { error = "Không thể lấy thông tin người dùng" });
        }
    }

    [HttpPost]
    public async Task<IActionResult> CreateUser([FromBody] CreateUserRequest request)
    {
        try
        {
            var id = Guid.NewGuid().ToString();
            var createdAt = DataHelpers.ToMySQLDateTime();

            // Hash password
            string passwordHash;
            if (!string.IsNullOrEmpty(request.Password))
            {
                passwordHash = BCrypt.Net.BCrypt.HashPassword(request.Password, 10);
            }
            else
            {
                return BadRequest(new { error = "Vui lòng nhập mật khẩu" });
            }

            // Validate role
            string? roleId = request.Role;
            if (string.IsNullOrEmpty(roleId))
            {
                var defaultRole = await _db.QueryFirstOrDefaultAsync<dynamic>("SELECT id FROM roles LIMIT 1");
                if (defaultRole == null)
                {
                    return BadRequest(new { error = "No roles found in database. Please create at least one role first." });
                }
                roleId = defaultRole.id?.ToString();
            }
            else
            {
                var roleCheck = await _db.QueryFirstOrDefaultAsync<dynamic>(
                    "SELECT id FROM roles WHERE id = ?", new { roleId });
                if (roleCheck == null)
                {
                    return BadRequest(new { error = $"Role with id {roleId} does not exist" });
                }
            }

            await _db.ExecuteAsync(
                @"INSERT INTO users (
                    id, name, email, phone, password_hash, role, status, avatar, created_at, updated_at
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
                new
                {
                    id,
                    name = request.Name,
                    email = request.Email,
                    phone = request.Phone ?? "",
                    passwordHash,
                    role = roleId,
                    status = request.Status,
                    avatar = DataHelpers.NormalizeString(request.Avatar),
                    createdAt,
                    updatedAt = createdAt
                });

            var newUser = await _db.QueryFirstOrDefaultAsync<dynamic>(
                @"SELECT 
                    u.id, u.name, u.email, u.phone, 
                    u.role,
                    COALESCE(r.description, r.name) as role_description,
                    u.status, u.avatar, u.created_at, u.updated_at 
                FROM users u
                LEFT JOIN roles r ON u.role = r.id
                WHERE u.id = ?",
                new { id });

            return StatusCode(201, newUser);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error creating user");
            if (ex.Message.Contains("Duplicate") || ex.Message.Contains("ER_DUP_ENTRY"))
            {
                return BadRequest(new { error = "Email đã tồn tại" });
            }
            return StatusCode(500, new { error = "Không thể tạo người dùng", message = ex.Message });
        }
    }

    [HttpPut("{id}")]
    public async Task<IActionResult> UpdateUser(string id, [FromBody] UpdateUserRequest request)
    {
        try
        {
            var existing = await _db.QueryFirstOrDefaultAsync<dynamic>(
                "SELECT * FROM users WHERE id = ?", new { id });

            if (existing == null)
            {
                return NotFound(new { error = "Không tìm thấy người dùng" });
            }

            var updates = new List<string>();
            var values = new Dictionary<string, object?>();

            if (request.Name != null)
            {
                updates.Add("name = ?");
                values["name"] = request.Name;
            }
            if (request.Email != null)
            {
                updates.Add("email = ?");
                values["email"] = request.Email;
            }
            if (request.Phone != null)
            {
                updates.Add("phone = ?");
                values["phone"] = request.Phone;
            }
            if (!string.IsNullOrEmpty(request.Password))
            {
                updates.Add("password_hash = ?");
                values["passwordHash"] = BCrypt.Net.BCrypt.HashPassword(request.Password, 10);
            }
            if (request.Role != null)
            {
                var roleCheck = await _db.QueryFirstOrDefaultAsync<dynamic>(
                    "SELECT id FROM roles WHERE id = ?", new { roleId = request.Role });
                if (roleCheck == null)
                {
                    return BadRequest(new { error = $"Role with id {request.Role} does not exist" });
                }
                updates.Add("role = ?");
                values["role"] = request.Role;
            }
            if (request.Status != null)
            {
                updates.Add("status = ?");
                values["status"] = request.Status;
            }
            if (request.Avatar != null)
            {
                updates.Add("avatar = ?");
                values["avatar"] = DataHelpers.NormalizeString(request.Avatar);
            }

            updates.Add("updated_at = ?");
            values["updatedAt"] = DataHelpers.ToMySQLDateTime();
            values["id"] = id;

            if (updates.Count > 1) // More than just updated_at
            {
                var sql = $"UPDATE users SET {string.Join(", ", updates)} WHERE id = ?";
                await _db.ExecuteAsync(sql, values);
            }

            var updated = await _db.QueryFirstOrDefaultAsync<dynamic>(
                @"SELECT 
                    u.id, u.name, u.email, u.phone, 
                    u.role,
                    COALESCE(r.description, r.name) as role_description,
                    u.status, u.avatar, u.created_at, u.updated_at 
                FROM users u
                LEFT JOIN roles r ON u.role = r.id
                WHERE u.id = ?",
                new { id });

            return Ok(updated);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error updating user");
            if (ex.Message.Contains("Duplicate") || ex.Message.Contains("ER_DUP_ENTRY"))
            {
                return BadRequest(new { error = "Email đã tồn tại" });
            }
            return StatusCode(500, new { error = "Không thể cập nhật người dùng", message = ex.Message });
        }
    }

    [HttpDelete("{id}")]
    public async Task<IActionResult> DeleteUser(string id)
    {
        try
        {
            var existing = await _db.QueryFirstOrDefaultAsync<dynamic>(
                "SELECT * FROM users WHERE id = ?", new { id });

            if (existing == null)
            {
                return NotFound(new { error = "Không tìm thấy người dùng" });
            }

            await _db.ExecuteAsync("DELETE FROM users WHERE id = ?", new { id });

            return NoContent();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error deleting user");
            return StatusCode(500, new { error = "Không thể xóa người dùng" });
        }
    }

    [HttpGet("me")]
    [Authorize]
    public async Task<IActionResult> GetCurrentUser()
    {
        try
        {
            var userId = HttpContext.GetUserId();
            if (string.IsNullOrEmpty(userId))
            {
                return Unauthorized(new { error = "Không có quyền truy cập" });
            }

            var sql = @"SELECT 
                u.id, u.name, u.email, u.phone, 
                u.role,
                COALESCE(r.description, r.name) as role_description,
                u.status, u.avatar, u.created_at, u.updated_at 
            FROM users u
            LEFT JOIN roles r ON u.role = r.id
            WHERE u.id = ?";

            var user = await _db.QueryFirstOrDefaultAsync<dynamic>(sql, new { userId });

            if (user == null)
            {
                return NotFound(new { error = "Không tìm thấy người dùng" });
            }

            // Convert avatar filename to URL if exists
            if (user.avatar != null)
            {
                user.avatar = FileUploadHelper.GetAvatarUrl(user.avatar?.ToString(), _configuration);
            }

            return Ok(user);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching current user");
            return StatusCode(500, new { error = "Không thể lấy thông tin người dùng hiện tại" });
        }
    }

    [HttpPost("me/avatar")]
    [Authorize]
    public async Task<IActionResult> UploadAvatar(IFormFile file)
    {
        try
        {
            var userId = HttpContext.GetUserId();
            if (string.IsNullOrEmpty(userId))
            {
                return Unauthorized(new { error = "Không có quyền truy cập" });
            }

            if (file == null || file.Length == 0)
            {
                return BadRequest(new { error = "Không có file được tải lên" });
            }

            if (!FileUploadHelper.IsImageFile(file))
            {
                return BadRequest(new { error = "Chỉ cho phép upload file ảnh (JPEG, JPG, PNG, GIF, WEBP)" });
            }

            if (file.Length > 2 * 1024 * 1024) // 2MB
            {
                return BadRequest(new { error = "File quá lớn. Kích thước tối đa là 2MB" });
            }

            var updatedAt = DataHelpers.ToMySQLDateTime();

            // Handle file upload
            var (filename, url) = await FileUploadHelper.HandleFileUpload(
                file, "avatars", _configuration, _environment);

            // Store filename or full URL in database
            var avatarValue = url.StartsWith("http") ? url : filename;

            // Update user avatar in database
            await _db.ExecuteAsync(
                "UPDATE users SET avatar = ?, updated_at = ? WHERE id = ?",
                new { avatar = avatarValue, updatedAt, userId });

            // Get updated user
            var sql = @"SELECT 
                u.id, u.name, u.email, u.phone, 
                u.role,
                COALESCE(r.description, r.name) as role_description,
                u.status, u.avatar, u.created_at, u.updated_at 
            FROM users u
            LEFT JOIN roles r ON u.role = r.id
            WHERE u.id = ?";

            var user = await _db.QueryFirstOrDefaultAsync<dynamic>(sql, new { userId });

            if (user == null)
            {
                return NotFound(new { error = "Không tìm thấy người dùng" });
            }

            user.avatar = FileUploadHelper.GetAvatarUrl(user.avatar?.ToString(), _configuration);

            return Ok(new
            {
                message = "Tải lên ảnh đại diện thành công",
                avatar = user.avatar,
                user
            });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error uploading avatar");
            return StatusCode(500, new { error = "Không thể tải lên ảnh đại diện", message = ex.Message });
        }
    }
}

